README.md
# Natural Language Date/Time processing #

## Some of the things you can do with nldt ##

### Get the weekday of a given date

    >>> import nldt
    >>> date = nldt.moment("2005-08-04")
    >>> date("%A")
    'Thursday'

### Find the first Sunday in a given month

    >>> import nldt
    >>> base = nldt.moment("2017-09-01")
    >>> wday = int(base("%u"))
    >>> delta = (6 - wday) * 24 * 3600
    >>> target = base + delta
    >>> target("%a %F")
    'Sat 2017-09-02'

### Find the date and weekday N days after or before Y

    >>> import nldt
    >>> Y = nldt.moment("1975-01-23")
    >>> X = Y - 200 * 24 * 3600
    >>> X("%a %F")
    'Sun 1974-07-07'
    >>> Z = Y + 75 * 24 * 3600
    >>> Z("%a %F")
    'Tue 1975-04-08'

### Parse date and time related natural langugae expressions

    >>> import nldt
    >>> P = nldt.Parser()
    >>> m = P('tomorrow')
    >>> str(m)
    '2018-02-28 20:58:01'
    >>> str(P('three days ago'))
    '2018-02-24 20:59:13'
    >>> str(P('next week'))
    '2018-03-05 00:00:00'
    >>> str(P('seven days from now'))
    '2018-03-06 21:18:07 Tue'

## How to think about times and timezones

When the input value of a moment is a single number, it is always
interpreted as the number of seconds since the epoch (on Unix, this is
1970-01-01T00:00:00 UTC) with no timezone adjustment. Such values are
always UTC times.

When the input value of a moment is a time.struct_time, a date/time tuple,
or a date/time specified in a string, a timezone adjustment is applied to
convert the local input time to UTC. By default, the local timezone is used
to interpret the input time, but you can specify another timezone to use
with the 'itz' argument to the moment constructor.

The function nldt.default_tz() can be used to change the default input
timezone.

## Classes ##

This module provides several classes intended to make date and time
functionality easy to access and use.

 * duration: Represents a length of time. Stored as a number of seconds.
   Can be used to represent the offset between UTC and a given timezone.

 * local: Holds information about the local timezone.

 * moment: Represents a point in time. Always stores its time reference in
   UTC. Can report its stored time in the format and/or timezone requested.

 * month: Holds information about months.

 * Parser: A natural language parser that can convert natural language
   expressions like "first week in January" or "last week" or "next year"
   into moments or durations.

### duration ###

 * The sum or difference of two durations is another duration. The sum of a
   moment and a duration is another moment. The difference of a moment and
   duration is another moment. The difference of two moments is a duration.
   It is not clear what the meaning of a sum of two moments would be, so
   that is undefined.

    * <duration> + <duration> -> <duration>
    * <duration> - <duration> -> <duration>
    * <moment> + <duration> -> <moment>
    * <moment> - <duration> -> <moment>
    * <moment> - <moment> -> <duration>
    * <moment> + <moment> -> UNDEFINED

### local ###
### moment ###

The constructor for the moment class can take input times in several forms.
Any input time that consists of a number or a string containing a number
will be interpreted as an epoch time, which is always in the UTC context.
That is, no timezone offset will be applied to such inputs.

Input times represented as a time.struct_tm, a tuple of 6 to 9 numbers, or
a date/time string will be interpreted in the default timezone unless an
explicit timezone is provided at construction time.

#### class moment

##### __init__(self, dt=None, fmt=None, itz=None)

The moment constructor takes the following (optional) arguments

 * dt: A date/time specification, which may be in any of a number of forms:
    * a number
    * a string containing all numeric digits
    * an already instantiated moment
    * a time.struct_time structure
    * a tuple containing between 6 and 9 numbers
    * a string containing a date/time specification in a recognized format
 * fmt: The format to use in interpreting a dt string
 * itz: The timezone used in interpreting the input time in some cases (see
   below)

A moment's internal time is always stored in UTC. Epoch times defined as
the number of seconds since the epoch always indicate a UTC time. If dt is
None, the current UTC time will be stored in the moment.

If dt is a number, a numeric string, or an already instantiated moment, the
new moment will store the UTC epoch time represented by the number, the number
represented by the string, or value from the already instantiated moment.
In all these cases, no timezone adjustment is made.

If dt is a time.struct_time, a tuple, or a date/time string, by default the
indicated time will be interpreted in terms of the local time zone and
adjusted to UTC before being stored in the moment. So, for example, if the
input time is "2011-01-01 00:00:00" and the local timezone is US/Eastern,
the moment will store 1293858000, which represents UTC time "2011-01-01
05:00:00".

The input timezone can be changed by passing the itz argument or by calling
moment.default_tz() before object instantiation.

##### default_tz(cls, itz=None)

This class method (i.e., it must be called as 'nldt.moment.default_tz()',
not as a method on an object) sets the default timezone for interpreting
input times when instantianting moments.

##### takes_tz(cls, dt)

This class method returns True if a timezone argument is valid when passing
dt to the moment constructor (i.e., dt is NOT a number, numeric string, or
already instantiated moment).

##### __call__(self, format=None, otz=None)

This method is normally called by appending parentheses to an instantiated
moment object. For example,

    import nlde
    q = nldt.moment()  # record the current time (converts local to UTC)
    q("%F %T")         # report current local time as YYYY-mm-dd HH:MM:SS
    q("%F %T", otz='Cuba')   # report current time in Cuba's timezone


### month ###
### Parser ###


The module will also includes a command line script for command line access
to the functionality provided by the library.

Internally, moments and durations are stored as a floating point number of
seconds. Points in time (moment instances) are stored as the number of
seconds since midnight at the beginning of January 1, 1970 (defined in
UTC). Durations are stored as the number of seconds in the time interval.

## Quick Start ##

    import nldt
    from nldt import moment, duration

Query the timezone configured for the current location

    >>> nldt.timezone()
    'EST'

Check whether Daylight Savings Time is in force now in the currently
configured timezone

    >>> nldt.dst()
    False

Check whether Daylight Savings Time is in force on a given date in a given
timezone. If timezone is not specified, the one configured for the current
location will be used. Both dst() and timezone() can take a moment object
(representing a UTC time) as an argument and will return their result in
terms of that moment in time.

    >>> nldt.dst(moment('2016-07-01'), tz='US/Eastern')
    True
    >>> nldt.dst(moment('2016-12-01'), tz='US/Central')
    False

Note that the short timezone name will vary with whether DST is in force.
For example, on 2016-07-01,

    >>> nldt.timezone(moment('2016-07-01'))
    'EDT'
    >>> nldt.timezone(moment('2016-12-31'))
    'EST'

Get the curent UTC date and time. The __call__ method on a moment object
(i.e., '<obj>()') returns the date stored by the object. By default, the
date is returned in ISO-8601 format (YYYY-mm-dd).

    >>> now = moment()
    >>> now()
    2016-11-30

To get the epoch time,

    >>> now.epoch()
    1480482000

To get the date and time in some other format, simply pass the desired
format as an argument:

    >>> now('%Y-%m-%d %H:%M:%S')
    2016-11-30 08:22:38

To get a local date and time, a timezone must be provided

    >>> now('%Y-%m-%d %H:%M:%S', tz='us/central')
    2016-11-30 02:22:28

The special timezone, 'local', can be used to get the time at the currently
configured location

    >>> now('%Y-%m-%d %H:%M:%S %Z', tz='local')
    2016-11-30 03:22:28 EST

All the strftime format specifiers are available (see 'pydoc
time.strftime')

    >>> now("%b %d, %Y")
    'Nov 30, 2016'

The sum or difference of a moment and a duration is another moment.

    >>> now()
    '2016-07-01'
    >>> then = now + duration(day=1)
    >>> then()
    '2016-07-02'

The duration may also be specified as an integer number of seconds

    >>> later = now - 24 * 3600
    >>> later()
    '2016-06-30'

The difference of two moments is a duration, which is reported by default
in terms of seconds

    >>> d = then - later
    >>> d
    duration(172800)
    >>> d()
    172800

Durations can report themselves in terms of other units

    >>> d('day')
    2.0
    >>> d('hour')
    48.0


## Reference ##

This section will list and describe all the classes and all public methods
for each class.

An object representing the current moment:

    >>> now = nldt.moment()

Available public methods:

    >>> now()                # __call__(fmt=None)
    '2016-12-03'

The __call__() method with no arguments displays the stored moment as an
ISO date. With a format, the stored moment is formatted according to the
argument:

    >>> now('%D %T')
    '12/03/16 12:27:49'

The __repr__() method produces a string that can be eval'd to unambiguously
generate an identical object:

    >>> a
    nldt.moment(1480786154)

The __str__() method produces a human readable string

    >>> str(a)
    '2016-12-03 12:29:14'

The __eq__() method is defined so that two moment objects based on the same
time index are considered equal.

    >>> a = nldt.moment('2016-12-02')
    >>> b = nldt.moment('yesterday')
    >>> a == b
    True

The dst() method returns True or False to indicate whether Daylight Savings
Time is in force.

    >>> yes = nldt.moment('2016-07-01')
    >>> yes.dst()
    True
    >>> no = nldt.moment('2016-12-01')
    >>> no.dst()
    False

The epoch() method returns the stored moment as an epoch time.

    >>> yes.epoch()
    1467345600

The localtime() method returns the tm tuple for the stored moment.

    >>> yes.localtime()
    time.struct_time(tm_year=2016, tm_mon=7, tm_mday=1,
                     tm_hour=0, tm_min=0, tm_sec=0,
                     tm_wday=4, tm_yday=183, tm_isdst=1)

The timezone() method returns the configured timezone based on whether the
stored time indicates DST is in force.

    >>> yes.timezone()
    'EDT'
    >>> no.timezone()
    'EST'

The parse() method accepts a string argument containing a date/time
specification or a relative time expression in natural language and updates
the stored stored moment to match.

    >>> foo = nldt.moment()
    >>> str(foo)
    '2016-12-03 12:46:38'
    >>> foo.parse('next week')
    >>> str(foo)
    '2016-12-05 00:00:00'
    >>> foo.parse('last friday')
    >>> str(foo)
    '2016-12-02 00:00:00'
    >>> foo.parse('next thursday')
    >>> str(foo)
    '2016-12-08 00:00:00'


## More Tricks ##

The nldt module will try to intuit the format of dates passed to it

    >>> later = nldt.moment('Dec 29 2016')
    >>> later()
    '2016-12-29'

However, it will also accept a format specification to guide its parsing
(but in this case, the date spec must match the format)

    >>> later = nldt.moment('Dec 29 2016', "%b %d %Y")
    >>> later()
    '2016-12-29'

    >>> later = nldt.moment('dec 29 2016', '%b %d, %Y')
    Traceback (most recent call last):
     ...
    ValueError: time data 'dec 29 2016' does not match format '%b %d, %Y'

Offsets can also be specified as arguments...

    >>> then = nldt.moment("tomorrow")
    >>> then("%Y-%m-%d")
    '2016-12-01'

... or natural language expressions (these can also be passed to the
constructor)

    >>> now.parse("last saturday")
    >>> now()
    '2016-11-26 00:15:30'

    >>> then.parse("fri week")
    >>> then()
    '2016-12-09 00:15:45'

    >>> now.parse('next month')
    >>> now.parse('next year')
    >>> now.parse('last week')
    >>> now.parse('last month')
    >>> now.parse('last year')
    >>> now.parse('a week ago wednesday')                       !@!
    >>> now.parse('1 month 2 days 3 weeks 4 minutes from sunday')                       !@!
    >>> now.parse('42 hrs 42 seconds ago')                       !@!
    >>> now.parse('now+ 42 days 42 weeks')                       !@!
    >>> now.parse('jan 12')                       !@!
    >>> now.parse('jUlY 28 1996 23:59:59')                       !@!
    >>> now.parse('28th')                       !@!
    >>> now.parse('February 7')                       !@!
    >>> now.parse('mar 4th')                       !@!
    >>> now.parse('next april 3rd')                       !@!
    >>> now.parse('last sep-19')                       !@!
    >>> now.parse('last sep-19 7pm')                       !@!
    >>> now.parse('next july 10th at 19:00')                       !@!
    >>> now.parse('7:30 on the 4th of next month')                       !@!
    >>> now.parse('fifteenth of last month')                       !@!
    >>> now.parse('22nd October 3:42am')                       !@!
    >>> now.parse('31st of dec at 9:30 pm')                       !@!
    >>> now.parse('Tue Apr 03 17:00:00 CST 2007')                       !@!
    >>> now.parse('June 13, 1997')                       !@!
    >>> now.parse('May 5,2001')                       !@!
    >>> now.parse('feb 3 1980')                       !@!
    >>> now.parse('october 7 1993 at midnight')                       !@!
    >>> now.parse('01/01/01')                       !@!
    >>> now.parse('3/4/5')                       !@!
    >>> now.parse('30/2/2000')                       !@!
    >>> now.parse('2/30/01')                       !@!
    >>> now.parse('11/31/05')                       !@!

All the time functions are supported:

    now = when.when()
    now.time()                # same as now.epoch()           !@!
    now.clock()               # simply calls time.clock()     !@!
    now.sleep()               # calls time.sleep()            !@!
    now.gmtime()                                              !@!
    now.localtime()                                           !@!
    now.asctime()                                             !@!
    now.ctime()                                              !@!
    now.mktime()              # same as now.epoch()          !@!
    now.strftime(fmt)         # format stored time with *fmt*   !@!
    now.strptime(spec, fmt)   # update stored time to *spec* per *fmt*  !@!
    now.tzset(timezone)       # convert stored time to new timezone  !@!

Taking the difference of two points in time produces a duration object

    >>> delta = then - now            !@!
    >>> delta.seconds()               !@!
    1123133

    >>> delta.days()                  !@!
    12.999224537037037

A duration added to a when object produces another when object

    >>> later = then + delta         !@!
    >>> later()
    '2016-12-21 00:31:19'

In the strftime specification, '%u' generates the 1-based weekday number
where Monday is the first day of the week.

In the strftime specification, '%w' generates the 0-based weekday number
where Sunday is the first day of the week.

The strftime(3) man page says that in the tm structure, tm_wday is the day
of the week with Sunday == 0.

The python time.localtime() function returns a tm structure with tm_wday
being 0 for Monday through 6 for Sunday.


    now = nldt.moment()
    then = now
    then.update(tomorrow)
    diff = then - now


    now.update('next week')
    now.update('second wednesday of next july')
    now.yesterday()
    now.last_week()
    now.last_month()
    now.parse('end of last month')
    now.last_month('<')
    now.last_month('>')
    now.last_month(

## Contributor Notes ##

 * Functions, methods, and variables whose names begin with alphabetic
   letters and show up in the pydoc output are part of the public API. When
   such routines return a moment or duration, by default they should do so
   in a human-readable format.

 * The prefered human-readable format is the ISO format: %Y-%m-%d
   [%H:%M:%S]

 * Internal functions, methods, and variables, whose names begin with a
   single underscore, generally should expect and return integer epoch
   times.


CHANGELOG.md
## 0.0.20 ... 2019-10-14 13:30:35

 * Protect escape sequence strings with a prepended 'r'.
 * Copy of regexp in a comment was breaking flake.
 * Comment tests based on timezone 'Pacific/Norfolk' since it's not
   functional on TravisCI.

## 0.0.19 ... 2019-10-14 10:36:35

 * On travis CI, only build branch travis or versioned releases.

## 0.0.18 ... 2018-03-18 16:54:26

 * 'nldt now' reports local time, not UTC.
 * Reorganized tests to represent the layer cake.
 * Obsolete files test_nldt.py, test_cmdl.py, etc.
 * Class 'local' renamed to 'timezone'.
 * Consolidated tests for nldt.dst().
 * Given no argument, nldt.dst() reports about current time.
 * Prefer hyphens over underscores where possible.
 * Add test-only extension to text catalog.
 * Simplify generation of expected time values in tests.

## 0.0.17

 * Ensure that the default timezone is local for both input (when
   constructing moments) and output (when reporting out the stored time).
   The default timezone can be overridden for input by calling
   moment.default_tz() or by passing the itz argument to the moment
   constructor. The default can be overridden for output by using 'with
   nldt.tz_context' or passing the otz argument to moment.__call__().
 * Put message strings in text.py and use them from there.
 * Move 'class' away from the beginning of docstrings since otherwise emacs
   python-mode indentation is messed up.
 * Ability to reset default timezone to the factory default.
 * Ability for week start to be arbitrary day.
 * Full test coverage for moment.__eq__(), nldt.tzstring(), nldt.isnum().
 * Make test case ids more descriptive.
 * Get tz_context() working for all timezones.
 * Test case consolidation, code cleanup.

## 0.0.16

 * Renamed the 'tz' argument to moment.__init__() to 'itz' to indicate that
   the timezone is applied to the *input* date/time spec
 * Renamed the 'tz' argument to moment.__call__() to 'otz' to indicate that
   the timezone is applied to the moment on *output*
 * Add test for nldt.moment.takes_tz() to ensure 100% test coverage.
 * Add a test that tickles the bug where 'with nldt.tz_context()' throws a
   KeyError on exit when 'TZ' is not in the environment. Update the code to
   satisfy the test (thereby fixing the bug).
 * Use nldt.moment.default_tz() to set the default input timezone for tests
   to 'utc' so that it doesn't have to be specified on every moment call.
 * Move the tests for nldt.moment.default_tz() to the end of test_moment.py
   so they don't interfere with other tests.
 * Adding reference info to README.md.

## 0.0.15

 * For duration constructor, interpret moment values in the 'utc' timezone
   context
 * New moment class method takes_tz() to indicate whether a moment arg can
   take a timezone argument or not
 * Adjust tests to assess difference thresholds rather than strict equality
   in some cases
 * Adjust time anchors for correctness and easier maintenance of tests
 * For now, we need explicit utc timezone context specification because we
   don't (yet) have a way of specifying a default timezone context for
   moment construction
 * Update tests to use monotonically increasing numeric test ids
 * Test and code to support moment subtraction
 * Test and code for nldt.moment.default_tz() to set default timezone for
   moment constructor input

## 0.0.14

 * Add test for methods ctime(), asctime() and code to satisfy the test
 * Expand test for nldt.localtime() and update code to satisfy it
 * Shorten test names to fit 80 column line

## 0.0.13

 * Add moment.time() as an alias for alias moment.epoch() (to parallel
   time.time())
 * Add nldt.clock() as an alias for time.clock()
 * Add nldt.duration.sleep() as an alias for time.sleep()

## 0.0.12

 * Update requirements.txt to include pexpect
 * Move numberize.py into nldt to make travis happy

## 0.0.11

 * Add tests for 'nldt now' with and without -z and -f
 * Add copyright/licensing information
 * Add object nldt.object and a test for it
 * Add 'with nldt.timezone(TZNAME)' context manager (had to rename previous
   nldt.timezone function to nldt.localzone)
 * Reordered source code to make it easier to find stuff

## 0.0.10 ...

 * Add nldt command line program, handle 'now', 'today', 'tomorrow',
   'yesterday'
 * Adjust command line argument handling to interpret no arguments as 'now'
 * Fix bug where time.mktime() assumes its input tuple is local time by
   calling tzset() to adjust the local timezone when necessary
 * Fix bug where moment was not initializing itself from a numeric string
 * Fix a bug where no parse branch matches the input
 * In tests, avoid the assumption that 'local' always means 'US/Eastern'

## 0.0.9 ... 2018.0212 10:08:35

 * Move the authoritative project version from ~/version.py to
   ~/nldt/verinfo.py. Add function nldt.version() which returns the current
   project version. Test test_version() verifies that nldt.version()
   returns a 'reasonable' version value (i.e., three dot separated
   numbers). In test_deployable(), the version string is checked against
   the most recent git tag.
 * Make duration objects callable with an optional format, like moment
   objects. Remove obsoleted nldt.hhmm(), test_helpers.py,
   duration.format().

## 0.0.8 ... 2018.0211 18:05:56

 * Make nldt.dst() work for timezones that don't have a transition table in
   the pytz tzinfo object.
 * Travis testing apparently takes place in the UTC timezone.
 * nldt.dst() now looks through the pytz _utc_transition_times table to
   decide whether the DST flag should be on or off for a given timezone at
   a specified moment in time.

## 0.0.7 ... 2018.0211 11:07:35

 * Debugging Travis support -- flake8 has to be installed

## 0.0.6 ... 2018.0211 10:49:56

 * Extended flake8 coverage to all the python files
 * Add travis support

## 0.0.5 ... 2018.0210 13:12:53

 * Check tests for debuggability
 * Compare moments for equality based on int epoch, not float moment

nldt/text.py
"""
nldt - Natural Language Date/Time support
Copyright (c) 2017 - <the end of time>  Tom Barron
See file LICENSING for details
-------------------------------------------------------------------------------
This module defines strings used in nldt to minimize repetition (DRY).

txt['err-*']      Error messages
txt['exc-*']      Exception names
txt['tz-*']       Timezone names
txt['xpr-*']      Time expressions
txt['*-f']        String contains format spec ('{}')
"""
txt = {}
txt['ABC-noinst'] = "This is an abstract base class -- don't instantiate it."
txt['arg-more'] = "argument must be moment or epoch number"
txt['date01'] = "2010-01-01"
txt['date02'] = "2000-12-31 15:59:59"
txt['date03'] = "2010-12-31"
txt['2010-end'] = "2010-12-31"
txt['2011-begin'] = "2011-01-01"
txt['date04'] = "2012-07-01"
txt['leap-yester'] = "2012-02-28"
txt['leap-today'] = "2012-02-29"

txt['dst-when'] = "dst() 'when' argument must be str, number, or moment"
txt['epc-nofmttz'] = "moment(epoch) does not take timezone or format"

txt['err-abc'] = "This is an abstract base class -- don't instantiate it."
txt['err-argmore'] = "argument must be moment or epoch number"
txt['err-ceilweek'] = "ceiling() only accepts start for unit='week'"
txt['err-notatime'] = ("Failure parsing '{}' -- not recognized as a"
                       " time expression")
txt['err-nounit'] = "No unit found in expression"
txt['err-nounit-f'] = "No unit found in expression '{}'"
txt['err-indxfy'] = "Could not indexify '{}'"
txt['err-indxfy-f'] = "Could not indexify '{}'"
txt['err-indxfy-nof'] = "Could not indexify"

txt['exc-ulerr'] = "UnboundLocalError"

txt['fmt-str'] = "moment() cannot take format when date is not of type str"
txt['inv-subtrahend'] = "Invalid subtrahend for moment subtraction"
txt['invtup'] = "Invalid tm tuple"
txt['iso-date'] = "%Y-%m-%d"
txt['iso-ymdhms'] = "%Y.%m%d %H:%M:%S"
txt['iso-datetime'] = "%Y-%m-%d %H:%M:%S"
txt['mctor-001'] = "If start or end is specified, both must be"
txt['mom-sum'] = "sum of moments is not defined"
txt['nan'] = "not a number"
txt['no-args'] = "moment() cannot take format or tz without date spec"
txt['no-match'] = ("None of the common specifications match"
                   " the date/time string")
txt['no-unit'] = "No unit found in expression '{}'"
txt['not-indxfy'] = "Could not indexify '{}'"
txt['not-empty'] = "result must be an empty list"
txt['not-timeu'] = "'{}' is not a time unit"
txt['optypes-01'] = "Unsupported operand types(s): try 'moment' - 'duration'"
txt['optypes-02'] = "unsupported operand types(s): '{}' and '{}'"
txt['parse-fail'] = ("Failure parsing '{}' -- not recognized as"
                     " a time expression")
txt['start-inv01'] = "start only valid in ceiling/floor when unit='week'"
txt['start-inv02'] = "start must be a weekday name or abbreviation"
txt['stubmsg'] = "{}() is a stub -- please complete it."
txt['tuplen'] = "need at least 6 values, no more than 9"
txt['tz-addis'] = "Africa/Addis_Ababa"
txt['tz-ak'] = "US/Alaska"
txt['tz-est'] = "US/Eastern"
txt['tz-nz'] = "NZ"
txt['utc-offset'] = "utc_offset requires an epoch time or None"
txt['valid-calls'] = "\n".join(["Valid ways of calling nldt.moment():",
                                "    nldt.moment()",
                                "    nldt.moment(<epoch-seconds>)",
                                "    nldt.moment('YYYY-mm-dd')",
                                "    nldt.moment(<date-str>[, <format>])"])
txt['wday-rgx'] = "(mon|tues|wednes|thurs|fri|satur|sun)day"

txt['xpr-4dotw'] = "fourth day of this week"
txt['xpr-5dolw'] = "fifth day of last week"
txt['xpr-lwk'] = "last week"
txt['xpr-nwk'] = "next week"
txt['xpr-nthu'] = "next thursday"
txt['xpr-nfri'] = "next friday"
txt['xpr-tod'] = "the other day"
